src/datahike/index/persistent_set.cljc

```diff
@@ -31,54 +24,103 @@
     :avet dd/cmp-datoms-avet-quick
     dd/cmp-datoms-eavt-quick))
 
-(defn -insert [set datom index-type]
-  (set/conj set datom (index-type->cmp-quick index-type)))
-
-(defn -remove [set datom index-type]
-  (set/disj set datom (index-type->cmp-quick index-type)))
-
-;; Functions defined in multimethods in index.cljc
-
-(defn empty-set [index-type]
-  (set/sorted-set-by (index-type->cmp index-type)))
-
-(defn init-set [datoms index-type indexed]
+(extend-type PersistentSortedSet
+  IIndex
+  (-all [pset]
+    (identity pset))
+  (-seq [pset]
+    (seq pset))
+  (-count [pset]
+    (count pset))
+  (-insert [pset datom index-type _op-count]
+    (-> (or (when-let [old (first (set/slice pset
+                                             (dd/datom (.-e datom) (.-a datom) (.-v datom) tx0)
+                                             (dd/datom (.-e datom) (.-a datom) (.-v datom) txmax)))]
+              (set/disj pset old (index-type->cmp-quick index-type)))
+            pset)
+        (set/conj datom (index-type->cmp-quick index-type))))
+  (-temporal-insert [pset datom index-type _op-count]
+    (set/conj pset datom (index-type->cmp-quick index-type)))
+  (-upsert [pset datom index-type _op-count]
+    (-> (or (when-let [old (first (set/slice pset
+                                             (dd/datom (.-e datom) (.-a datom) nil tx0)
+                                             (dd/datom (.-e datom) (.-a datom) nil txmax)))]
+              (set/disj pset old (index-type->cmp-quick index-type)))
+            pset)
+        (set/conj datom (index-type->cmp-quick index-type))))
+  (-temporal-upsert [pset datom index-type _op-count]
+    (-> (or (when-let [old (first (set/slice pset
+                                             (dd/datom (.-e datom) (.-a datom) nil tx0)
+                                             (dd/datom (.-e datom) (.-a datom) nil txmax)))]
+              (set/conj pset (dd/datom (.-e old) (.-a old) (.-v old) (.-tx old) false)
+                        (index-type->cmp-quick index-type)))
+            pset)
+        (set/conj datom (index-type->cmp-quick index-type))))
+  (-remove [pset datom index-type _op-count]
+    (set/disj pset datom (index-type->cmp-quick index-type)))
+  (-slice [pset from to _]
+    (set/slice pset from to))
+  (-flush [pset _]
+    (set! (.-_root pset)
+          (set/-flush (.-_root pset)))
+    pset)
+  (-transient [pset]
+    (transient pset))
+  (-persistent! [pset]
+    (persistent! pset)))
+
+(defn get-loader [konserve-store]
+  (proxy [Loader] []
+    (load [address]
+      (let [children-as-maps (async/<!! (k/get konserve-store address))]
```

Comment:
It is not good to use `<!!` here, it would be better to use synchronous konserve calls instead.

Reply:
As far as I know, this is not enabled yet for datahike. I added a note to change this later though.